x-pack/plugins/ml/public/application/util/timeseriesexplorer_utils.ts

```diff
@@ -0,0 +1,631 @@
+              selectedJob,
+              detectorIndex,
+              forecastId,
+              nonBlankEntities,
+              searchBounds.min.valueOf(),
+              searchBounds.max.valueOf(),
+              focusAggregationInterval.asMilliseconds(),
+              aggType
+            );
+          })()
+        : of(null),
+    ]).pipe(
+      map(
+        ([metricData, recordsForCriteria, scheduledEventsByBucket, annotations, forecastData]) => {
+          // Sort in descending time order before storing in scope.
+          const anomalyRecords = recordsForCriteria?.records
+            .sort((a, b) => a[TIME_FIELD_NAME] - b[TIME_FIELD_NAME])
+            .reverse();
+
+          const scheduledEvents = scheduledEventsByBucket?.events[selectedJob.job_id];
+
+          let focusChartData = processMetricPlotResults(metricData.results, modelPlotEnabled);
+          // Tell the results container directives to render the focus chart.
+          focusChartData = processDataForFocusAnomalies(
+            focusChartData,
+            anomalyRecords,
+            focusAggregationInterval,
+            modelPlotEnabled,
+            functionDescription
+          );
+          focusChartData = processScheduledEventsForChart(
+            focusChartData,
+            scheduledEvents,
+            focusAggregationInterval
+          );
+
+          const refreshFocusData: FocusData = {
+            scheduledEvents,
+            anomalyRecords,
+            focusChartData,
+          };
+
+          if (annotations) {
+            if (annotations.error !== undefined) {
+              refreshFocusData.focusAnnotationError = annotations.error;
+              refreshFocusData.focusAnnotationData = [];
+            } else {
+              refreshFocusData.focusAnnotationData = (
+                annotations.annotations[selectedJob.job_id] ?? []
+              )
+                .sort((a, b) => {
+                  return a.timestamp - b.timestamp;
+                })
+                .map((d, i: number) => {
+                  d.key = (i + 1).toString();
+                  return d;
+                });
+            }
+          }
+
+          if (forecastData) {
+            refreshFocusData.focusForecastData = processForecastResults(forecastData.results);
+            refreshFocusData.showForecastCheckbox = refreshFocusData.focusForecastData.length > 0;
+          }
+          return refreshFocusData;
+        }
+      )
+    );
+  }
+
+  return {
+    getAutoZoomDuration,
+    calculateAggregationInterval,
+    calculateInitialFocusRange,
+    calculateDefaultFocusRange,
+    processRecordScoreResults,
+    processMetricPlotResults,
+    processForecastResults,
+    findChartPointForAnomalyTime,
+    processDataForFocusAnomalies,
+    findChartPointForScheduledEvent,
+    processScheduledEventsForChart,
+    getFocusData,
+  };
+}
+
+export type TimeSeriesExplorerService = ReturnType<typeof timeSeriesExplorerProvider>;
```

Comment:
we should add a hook here which can be used in tsx files.
```ts
export function useTimeSeriesExplorerService(): TimeSeriesExplorerService {
  const {
    services: {
      uiSettings,
      mlServices: { mlApiServices },
    },
  } = useMlKibana();
  const mlResultsService = mlResultsServiceProvider(mlApiServices);

  const mlTimeSeriesExplorer = useMemo(
    () => timeSeriesExplorerProvider(uiSettings, mlApiServices, mlResultsService),
    [uiSettings, mlApiServices, mlResultsService]
  );
  return mlTimeSeriesExplorer;
}
```

Reply:
The same can be done for `timeBucketsProvider`

Reply:
Updated in f682daaa0b13b541a66cc5725072a61828502541